Implement mediasession set positon state

shnmorimoto opened the pull request:
<!-- Please describe your changes on the following line: -->

fix #24808 

> Bonus points if you want to tweak the existing UI by adding a progress bar, and the info about the current position and total duration.

I haven't implemented this yet.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24808 (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

shnmorimoto committed:
implement MediaSession.setPositionState method

shnmorimoto committed:
updated tests for mediasession

shnmorimoto committed:
extended MediaSessionEvent with SetPositionState

shnmorimoto committed:
fix typo

shnmorimoto committed:
extended HostTrait with on_media_session_set_position_state

shnmorimoto committed:
extended android jni with onMediaSessionSetPositionState

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/htmlmediaelement.rs, components/script/dom/mediasession.rs, components/script/dom/webidls/MediaSession.webidl
 * @kichjang: components/script/dom/htmlmediaelement.rs, components/script/dom/mediasession.rs, components/script/dom/webidls/MediaSession.webidl

shnmorimoto commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +196,76 @@ impl MediaSessionMethods for MediaSession {
+        let mut playback_rate: f64 = state.playbackRate.map(|pr| *pr).unwrap_or(0.0);
+
+        // If the state is an empty dictionary then clear the position state.
+        if state.duration.is_none() && state.position.is_none() && state.playbackRate.is_none() {
```
https://w3c.github.io/mediasession/#dom-mediasession-setpositionstate

> If the state is an empty dictionary then clear the position state.

Is the position state the combination of duration and playback_position and playbackRate in media_instance(type is HTMLMediaElement)?
And, Does clear mean to set at 0?

shnmorimoto replied:
Thanks! I added the reset method, and fixed to use it.

shnmorimoto commented:
https://github.com/servo/servo/issues/24808#issuecomment-557461228

> Extend the existing HostTrait trait with a new on_media_session_set_position_state or such method. And implement that method for the C and JNI APIs

> For the Android part, extend the existing Callbacks and Client interfaces with the new onMediaSessionSetPositionState callback. And implement it here and here.

To be honest, I don't have enough understanding of these behaviors.
Is there any way to test that these are implemented correctly?

jdm commented:
r? @ferjm

ferjm commented on ports/libsimpleservo/jniapi/src/lib.rs:
```diff
@@ -560,6 +560,31 @@ impl HostTrait for HostCallbacks {
+        let position = position.map(|v| JValue::Float(v as jfloat)).unwrap_or(null);
+        let playback_rate = playback_rate
+            .map(|v| JValue::Float(v as jfloat))
+            .unwrap_or(null);
```
These `unwrap_or(null)`s won't build as it is expecting a `JSValue`. Anyway, as I pointed out before, I think we should not use `Option`s here and always pass the updated value for all the `MediaPositionState` members.

shnmorimoto replied:
fixed it.

ferjm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1925,6 +1925,18 @@ impl HTMLMediaElement {
+        self.duration.set(duration);
+    }
+
+    pub fn set_playback_rate(&self, playbackRate: f64) {
```
Use the existing `SetPlaybackRate` instead, please.

shnmorimoto replied:
fixed it.

ferjm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1925,6 +1925,18 @@ impl HTMLMediaElement {
+        self.playbackRate.set(playbackRate);
+    }
+
+    pub fn set_playback_position(&self, playback_position: f64) {
```
Use the existing `SetCurrentTime` instead, please.

shnmorimoto replied:
fixed it.

ferjm commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +196,76 @@ impl MediaSessionMethods for MediaSession {
+        if let Some(media_instance) = self.media_instance.get() {
+            media_instance.set_duration(duration);
+            media_instance.set_playback_rate(position);
+            media_instance.set_playback_position(playback_rate);
```
Actually, use the suggested existing functions instead of `set_playback_rate` and `set_playback_position` please.

shnmorimoto replied:
I understand. I fixed to use this.

ferjm reviewed (changes requested):
This is a great start, but we need to make some changes.

In addition to the comments I added, we should send `MediaSessionEvent::SetPositionState` events when the media player [notifies about position updates](https://github.com/servo/servo/blob/769df62ff12514af19b9090db3cfc13acc5216c1/components/script/dom/htmlmediaelement.rs#L1775).

ferjm commented:
> To be honest, I don't have enough understanding of these behaviors.
Is there any way to test that these are implemented correctly?

You can test on Android. The instructions to build for these platform are on the [README](https://github.com/servo/servo#cross-compilation-for-android) and the [wiki](https://github.com/servo/servo/wiki/Building-for-Android).

You can try modifying [this](https://github.com/ferjm/web-api-tests/blob/gh-pages/video/index.html#L96) example.

shnmorimoto commented on components/embedder_traits/lib.rs:
```diff
@@ -240,11 +240,21 @@ pub enum MediaSessionPlaybackState {
+    pub duration: Option<f64>,
+    pub playback_rate: Option<f64>,
+    pub position: Option<f64>,
+}
```
Thanks, fixed it.

shnmorimoto commented on ports/libsimpleservo/api/src/lib.rs:
```diff
@@ -594,6 +601,14 @@ impl ServoGlue {
                             .callbacks
                             .host_callbacks
                             .on_media_session_playback_state_change(state as i32),
+                        MediaSessionEvent::SetPositionState(positionState) => self
```
Oops, sorry. fixed it.

shnmorimoto commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +196,76 @@ impl MediaSessionMethods for MediaSession {
+            }
+            if let Some(state_duration) = state.duration {
+                if *state_position > *state_duration {
+                    return Err(Error::Type("position is grater than duration".to_owned()));
```
Oops, I'm sorry for this too. fixed it.

shnmorimoto commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +196,76 @@ impl MediaSessionMethods for MediaSession {
+    fn SetPositionState(&self, state: &MediaPositionState) -> Fallible<()> {
+        let duration: f64 = state.duration.map(|d| *d).unwrap_or(0.0);
+        let mut position: f64 = state.position.map(|p| *p).unwrap_or(0.0);
+        let mut playback_rate: f64 = state.playbackRate.map(|pr| *pr).unwrap_or(0.0);
```
I understand! fixed it.

shnmorimoto commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +196,76 @@ impl MediaSessionMethods for MediaSession {
+        // If the position is not present or its value is null, set it to zero.
+        if state.position.is_none() {
+            position = 0.0;
+        }
```
deleted it.

shnmorimoto commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +196,76 @@ impl MediaSessionMethods for MediaSession {
+        // If the playbackRate is not present or its value is null, set it to 1.0.
+        if state.playbackRate.is_none() {
+            playback_rate = 1.0;
+        }
```
deleted it.

shnmorimoto commented on ports/libsimpleservo/api/src/lib.rs:
```diff
@@ -134,6 +134,13 @@ pub trait HostTrait {
     fn on_media_session_metadata(&self, title: String, artist: String, album: String);
     /// Called when the media sessoin playback state changes.
     fn on_media_session_playback_state_change(&self, state: i32);
+    /// Called when the media sessoin position state is set.
```
fixed it.

shnmorimoto commented on ports/libsimpleservo/jniapi/src/lib.rs:
```diff
@@ -560,6 +560,31 @@ impl HostTrait for HostCallbacks {
+            "on_media_session_playback_state_change ({:?}, {:?}, {:?})",
+            duration, position, playback_rate
+        );
+        let env = self.get_env().unwrap();
```
Thanks! fixed it.

shnmorimoto commented:
@ferjm Thank you for a great review. 

> In addition to the comments I added, we should send MediaSessionEvent::SetPositionState events when the media player notifies about position updates.

I added code for MediaSessionEvent::SetPositionState.

shnmorimoto commented:
> You can test on Android. The instructions to build for this platform are on the README and the wiki.
Thanks! I don't have Android right now, so I'll get it and try it.

shnmorimoto commented:
When I change the code under the ports(ex. `ports/libsimpleservo/jniapi/src/lib.rs` ), it seems that it is not build even if I run `./mach build --dev `.

Is this my misunderstanding?

jdm commented:
The build command only builds one of the ports at a time. For libsimpleservo/jniapi, you need to pass the `--android` argument to `./mach build` (https://github.com/servo/servo/wiki/Building-for-Android).

shnmorimoto commented on support/android/apk/servoapp/src/main/java/org/mozilla/servo/MainActivity.java:
```diff
@@ -255,4 +255,15 @@ public void onMediaSessionPlaybackStateChange(int state) {
+            mMediaSession = new MediaSession(mServoView, this, getApplicationContext());
+        }
+
+        mMediaSession.setPositionState(duration, position, playbackRate);
```
I have a wrong implementation.
Should I call `onMediaSessionSetPositionState` of `Servo.java` directly here?
Or is it better to implement `setPositionState ` in `MediaSession.java` and call it there?

ferjm replied:
`Servo.onMediaSessionSetPositionState` is what calls this method (note that `MainActivity` implements the `Servo.Client` interface).

We need to implement `MediaSession.setPositionState`. That's where we should modify the existing notification UI to add the current position, the total duration and a progress bar. If you don't feel like hacking on the Android UI, leave the method unimplemented.

shnmorimoto replied:
Thanks!
I don't have a good understanding of the Android UI.
So this time I will not implement it. Sorry.

shnmorimoto commented on support/android/apk/servoapp/src/main/java/org/mozilla/servo/MediaSession.java:
```diff
@@ -192,4 +192,8 @@ public void updateMetadata(String title, String artist, String album) {
+
+    // Not implemented
+    // see https://github.com/servo/servo/pull/24885#discussion_r352496117
+    public void setPositionState(float duration, float position, float playbackRate) {}
```
For compile, I added only the signature.

ferjm commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +197,64 @@ impl MediaSessionMethods for MediaSession {
+        // If the duration is not present or its value is null, throw a TypeError.
+        if state.duration.is_none() {
+            return Err(Error::Type(
+                "duration is not preset or its value is null".to_owned(),
```
typo: present

shnmorimoto replied:
Oops. This is also typo..., Sorry.
done.

ferjm commented on components/script/dom/mediasession.rs:
```diff
@@ -194,6 +197,64 @@ impl MediaSessionMethods for MediaSession {
+            }
+        }
+
+        // If the position is not present or its value is null, set it to zero.
```
Remove this comment, please.

shnmorimoto replied:
Thanks! done.

ferjm reviewed (approved):
Looks good. Thank you!

r=me with the last minor comments addressed.

ferjm commented:
Thanks!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 29d6a35 has been approved by `ferjm`

<!-- @bors-servo r=ferjm 29d6a35db7027ac75137379730c19ebd3e8a03b7 -->

bors-servo commented:
:hourglass: Testing commit 29d6a35db7027ac75137379730c19ebd3e8a03b7 with merge 27d3a462aeb8461c535112a9caf423bd3d307fba...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Hjlv9h-HTMafVkXDgrJuOw)

shnmorimoto commented:
In my local, build is success.
Should I rebase and put commits in one?

jdm commented:
`./mach build --libsimpleservo` fails with this error:
```
error[E0609]: no field `on_media_session_set_position_state` on type `CHostCallbacks`
   --> ports\libsimpleservo\capi\src\lib.rs:741:17
    |
741 |         (self.0.on_media_session_set_position_state)(duration, position, playback_rate);
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unknown field
    |
    = note: available fields are: `flush`, `make_current`, `on_alert`, `on_load_started`, `on_load_ended` ... and 11 others
error: aborting due to previous error
```

shnmorimoto commented:
@jdm Thanks! I'll check.

shnmorimoto commented:
I fixed an error when run `./mach build --libsimpleservo`.

```
> ./mach build --libsimpleservo                                                                                                                        implement_mediasession_set_positon_state
   Compiling simpleservo_capi v0.0.1 (/Users/shn/Work/servo/ports/libsimpleservo/capi)
   Completed simpleservo_capi v0.0.1 custom-build (run) in 1.2s
   Completed simpleservo_capi v0.0.1 in 195.3s
    Finished dev [unoptimized + debuginfo] target(s) in 3m 21s
Build Completed in 0:03:22
```

jdm commented:
@bors-servo r=ferjm

bors-servo commented:
:pushpin: Commit 41ff93e has been approved by `ferjm`

<!-- @bors-servo r=ferjm 41ff93eca2bf3e8a15d99fedbf7da3a7f9af638a -->

bors-servo commented:
:hourglass: Testing commit 41ff93eca2bf3e8a15d99fedbf7da3a7f9af638a with merge 4ba6bfa57331607a32d5e27c4f50c1422f6afdb9...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/aihj0FaETyWYq5KCFFTH4A)

jdm commented:
@bors-servo retry
* #25046

bors-servo commented:
:hourglass: Testing commit 41ff93eca2bf3e8a15d99fedbf7da3a7f9af638a with merge f31a88d85de2a7bdee65216b3b66b697b459a31d...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/dos-wzStSnmiXNJkA7YPDw)
Approved by: ferjm
Pushing f31a88d85de2a7bdee65216b3b66b697b459a31d to master...
